vignettes/namespaces.Rmd

```diff
@@ -0,0 +1,288 @@
+---
+title: "Hold Namespace For Me"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Hold Namespace For Me}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+```
+
+```{r}
+library("tinkr")
+library("magrittr")
+library("commonmark")
+library("xml2")
+library("xslt")
+library("purrr")
+```
+
+
+## XML namespaces
+
+XML namespaces are a lot like package namespaces in R: they allow you to avoid
+clashes of names [for example, table can represent data or furniture](https://www.w3schools.com/XML/xml_namespaces.asp). 
+
+By default, nodes in XML do not have namespaces unless you give them one, which
+means that when you use XPath search, you can use the node names by default:
+
+```{r}
+d <- xml2::read_xml("<document>
+    <paragraph>
+      <text>hello there</text>
+      <text> ello  here</text>
+    </paragraph>
+  </document>")
+xml2::xml_ns(d)
+xml2::xml_find_all(d, "//document")
+xml2::xml_find_all(d, "//text[contains(text(), 'hello')]")
+```
+
+However if there is a namespace added to a node, all of its descendants will 
+inherit the namespace, which affects your XPath expressions: 
+
+```{r}
+d <- xml2::read_xml("<document>
+    <paragraph xmlns='http://commonmark.org/xml/1.0'>
+      <text>hello there</text>
+      <text> ello  here</text>
+    </paragraph>
+  </document>")
+xml2::xml_ns(d)
+xml2::xml_find_all(d, "//document")
+xml2::xml_find_all(d, "//text[contains(text(), 'hello')]") # does not work
```

Comment:
Maybe instead of using code comments the code could be split and commented with text?

````suggestion
Using the same XPath query as before no longer works, our call to `xml2::xml_find_all()` returns nothing.
```{r}
xml2::xml_find_all(d, "//text[contains(text(), 'hello')]") 
````
When a namespace is specified with `xmlns=&lt;URI&gt;`, {xml2} assigns it a **default
amespace** prefix, which is `d1`.
Therefore editing our XPath query like so will work:

```{r}
xml2::xml_find_all(d, "//d1:text[contains(text(), 'hello')]")
```

But is it a good idea to use d1? No, the xml2 docs....